File: Samples/Basic/DocumentReader-sample_kotlin/app/src/main/java/com/regula/documentreader/fingerprintsample/BluetoothReaderHelper.java
package com.regula.documentreader.fingerprintsample;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;

public class BluetoothReaderHelper {

    public static void connectBluetooth(String mac, BluetoothAdapter mBluetoothAdapter, BluetoothReaderService mChatService){
        BluetoothDevice device = mBluetoothAdapter.getRemoteDevice(mac);
        mChatService.connect(device);
        Log.v("Connected device mac", device.getAddress());
    }

    /**
     * calculate the check sum of the byte[]
     *
     * @param buffer byte[] required for calculating
     * @param size   the size of the byte[]
     * @return the calculated check sum
     */
    public static int calcCheckSum(byte[] buffer, int size) {
        int sum = 0;
        for (int i = 0; i < size; i++) {
            sum = sum + buffer[i];
        }
        return (sum & 0x00ff);
    }

    /**
     * generate the fingerprint image
     *
     * @param data   image data
     * @param width  width of the image
     * @param height height of the image
     * @param offset default setting as 0
     * @return bitmap image data
     */
    public static byte[] getFingerprintImage(byte[] data, int width, int height, int offset) {
        if (data == null) {
            return null;
        }
        byte[] imageData = new byte[width * height];
        for (int i = 0; i < (width * height / 2); i++) {
            imageData[i * 2] = (byte) (data[i + offset] & 0xf0);
            imageData[i * 2 + 1] = (byte) (data[i + offset] << 4 & 0xf0);
        }
        byte[] bmpData = toBmpByte(width, height, imageData);
        return bmpData;
    }

    public static byte[] changeByte(int data) {
        byte b4 = (byte) ((data) >> 24);
        byte b3 = (byte) (((data) << 8) >> 24);
        byte b2 = (byte) (((data) << 16) >> 24);
        byte b1 = (byte) (((data) << 24) >> 24);
        byte[] bytes = {b1, b2, b3, b4};
        return bytes;
    }

    /**
     * generate the image data into Bitmap format
     *
     * @param width  width of the image
     * @param height height of the image
     * @param data   image data
     * @return bitmap image data
     */
    public static byte[] toBmpByte(int width, int height, byte[] data) {
        byte[] buffer = null;
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            DataOutputStream dos = new DataOutputStream(baos);

            int bfType = 0x424d;
            int bfSize = 54 + 1024 + width * height;
            int bfReserved1 = 0;
            int bfReserved2 = 0;
            int bfOffBits = 54 + 1024;

            dos.writeShort(bfType);
            dos.write(changeByte(bfSize), 0, 4);
            dos.write(changeByte(bfReserved1), 0, 2);
            dos.write(changeByte(bfReserved2), 0, 2);
            dos.write(changeByte(bfOffBits), 0, 4);

            int biSize = 40;
            int biWidth = width;
            int biHeight = height;
            int biPlanes = 1;
            int biBitcount = 8;
            int biCompression = 0;
            int biSizeImage = width * height;
            int biXPelsPerMeter = 0;
            int biYPelsPerMeter = 0;
            int biClrUsed = 256;
            int biClrImportant = 0;

            dos.write(changeByte(biSize), 0, 4);
            dos.write(changeByte(biWidth), 0, 4);
            dos.write(changeByte(biHeight), 0, 4);
            dos.write(changeByte(biPlanes), 0, 2);
            dos.write(changeByte(biBitcount), 0, 2);
            dos.write(changeByte(biCompression), 0, 4);
            dos.write(changeByte(biSizeImage), 0, 4);
            dos.write(changeByte(biXPelsPerMeter), 0, 4);
            dos.write(changeByte(biYPelsPerMeter), 0, 4);
            dos.write(changeByte(biClrUsed), 0, 4);
            dos.write(changeByte(biClrImportant), 0, 4);

            byte[] palatte = new byte[1024];
            for (int i = 0; i < 256; i++) {
                palatte[i * 4] = (byte) i;
                palatte[i * 4 + 1] = (byte) i;
                palatte[i * 4 + 2] = (byte) i;
                palatte[i * 4 + 3] = 0;
            }
            dos.write(palatte);

            dos.write(data);
            dos.flush();
            buffer = baos.toByteArray();
            dos.close();
            baos.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return buffer;
    }

    /**
     * method of copying the byte[] data with specific length
     *
     * @param dstbuf    byte[] for storing the copied data with specific length
     * @param dstoffset the starting point for storing
     * @param srcbuf    the source byte[] used for copying.
     * @param srcoffset the starting point for copying
     * @param size      the length required to copy
     */
    public static void memcpy(byte[] dstbuf, int dstoffset, byte[] srcbuf, int srcoffset, int size) {
        if (size >= 0) System.arraycopy(srcbuf, srcoffset, dstbuf, dstoffset, size);
    }

    public static byte[] checkHaveTis(byte[] mUpImage, byte[] mUpImage2, int mUpImageSize) {
        if ((mUpImage[0] == 'F') && (mUpImage[1] == 'T')) {
            memcpy(mUpImage2, 0, mUpImage, 20, mUpImageSize - 20);
        } else {
            memcpy(mUpImage2, 0, mUpImage, 0, mUpImageSize);
        }
        return mUpImage2;
    }

    public static StringBuilder addStatusListHex(byte[] data, int size) {
        StringBuilder text = new StringBuilder();
        for (int i = 0; i < size; i++) {
            text.append(" ").append(Integer.toHexString(data[i] & 0xFF).toUpperCase()).append("  ");
        }
        return text;
    }
}

File: Fingerprint/Fingerprint-kotlin/app/src/main/java/com/regula/fingerprint_kotlin/BluetoothReaderHelper.kt
package com.regula.fingerprint_kotlin

import android.bluetooth.BluetoothAdapter
import android.util.Log
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.util.*
import kotlin.experimental.and


object BluetoothReaderHelper {
    fun connectBluetooth(
        mac: String?,
        mBluetoothAdapter: BluetoothAdapter,
        mChatService: BluetoothReaderService
    ) {
        val device = mBluetoothAdapter.getRemoteDevice(mac)
        mChatService.connect(device)
        Log.v("Connected device mac", device.address)
    }

    /**
     * calculate the check sum of the byte[]
     *
     * @param buffer byte[] required for calculating
     * @param size   the size of the byte[]
     * @return the calculated check sum
     */
    fun calcCheckSum(buffer: ByteArray, size: Int): Int {
        var sum = 0
        for (i in 0 until size) {
            sum += buffer[i]
        }
        return sum and 0x00ff
    }

    /**
     * generate the fingerprint image
     *
     * @param data   image data
     * @param width  width of the image
     * @param height height of the image
     * @param offset default setting as 0
     * @return bitmap image data
     */
    fun getFingerprintImage(
        data: ByteArray?,
        width: Int,
        height: Int,
        offset: Int
    ): ByteArray? {
        if (data == null) {
            return null
        }
        val imageData = ByteArray(width * height)
        for (i in 0 until width * height / 2) {
            imageData[i * 2] = (data[i + offset] and 0xf0.toByte())
            imageData[i * 2 + 1] = (data[i + offset].toInt() shl 4 and 0xf0).toByte()
        }
        return toBmpByte(width, height, imageData)
    }

    private fun changeByte(data: Int): ByteArray {
        val b4 = (data shr 24).toByte()
        val b3 = (data shl 8 shr 24).toByte()
        val b2 = (data shl 16 shr 24).toByte()
        val b1 = (data shl 24 shr 24).toByte()
        return byteArrayOf(b1, b2, b3, b4)
    }

    /**
     * generate the image data into Bitmap format
     *
     * @param width  width of the image
     * @param height height of the image
     * @param data   image data
     * @return bitmap image data
     */
    private fun toBmpByte(width: Int, height: Int, data: ByteArray?): ByteArray? {
        var buffer: ByteArray? = null
        try {
            val baos = ByteArrayOutputStream()
            val dos = DataOutputStream(baos)
            val bfType = 0x424d
            val bfSize = 54 + 1024 + width * height
            val bfReserved1 = 0
            val bfReserved2 = 0
            val bfOffBits = 54 + 1024
            dos.writeShort(bfType)
            dos.write(changeByte(bfSize), 0, 4)
            dos.write(changeByte(bfReserved1), 0, 2)
            dos.write(changeByte(bfReserved2), 0, 2)
            dos.write(changeByte(bfOffBits), 0, 4)
            val biSize = 40
            val biPlanes = 1
            val biBitcount = 8
            val biCompression = 0
            val biSizeImage = width * height
            val biXPelsPerMeter = 0
            val biYPelsPerMeter = 0
            val biClrUsed = 256
            val biClrImportant = 0
            dos.write(changeByte(biSize), 0, 4)
            dos.write(changeByte(width), 0, 4)
            dos.write(changeByte(height), 0, 4)
            dos.write(changeByte(biPlanes), 0, 2)
            dos.write(changeByte(biBitcount), 0, 2)
            dos.write(changeByte(biCompression), 0, 4)
            dos.write(changeByte(biSizeImage), 0, 4)
            dos.write(changeByte(biXPelsPerMeter), 0, 4)
            dos.write(changeByte(biYPelsPerMeter), 0, 4)
            dos.write(changeByte(biClrUsed), 0, 4)
            dos.write(changeByte(biClrImportant), 0, 4)
            val palatte = ByteArray(1024)
            for (i in 0..255) {
                palatte[i * 4] = i.toByte()
                palatte[i * 4 + 1] = i.toByte()
                palatte[i * 4 + 2] = i.toByte()
                palatte[i * 4 + 3] = 0
            }
            dos.write(palatte)
            dos.write(data)
            dos.flush()
            buffer = baos.toByteArray()
            dos.close()
            baos.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return buffer
    }

    /**
     * method of copying the byte[] data with specific length
     *
     * @param dstbuf    byte[] for storing the copied data with specific length
     * @param dstoffset the starting point for storing
     * @param srcbuf    the source byte[] used for copying.
     * @param srcoffset the starting point for copying
     * @param size      the length required to copy
     */
    fun memcpy(dstbuf: ByteArray?, dstoffset: Int, srcbuf: ByteArray?, srcoffset: Int, size: Int) {
        if (size >= 0) System.arraycopy(srcbuf, srcoffset, dstbuf, dstoffset, size)
    }

    fun checkHaveTis(mUpImage: ByteArray, mUpImage2: ByteArray?, mUpImageSize: Int): ByteArray? {
        if (mUpImage[0] == 'F'.code.toByte()
            && mUpImage[1] == 'T'.code.toByte()
        ) {
            memcpy(mUpImage2, 0, mUpImage, 20, mUpImageSize - 20)
        } else {
            memcpy(mUpImage2, 0, mUpImage, 0, mUpImageSize)
        }
        return mUpImage2
    }

    fun addStatusListHex(data: ByteArray, size: Int): StringBuilder {
        val text = StringBuilder()
        for (i in 0 until size) {
            text.append(" ")
                .append(
                    Integer.toHexString(data[i].toInt() and 0xFF).uppercase(Locale.getDefault())
                )
                .append("  ")
        }
        return text
    }
}